lib/semian/protected_resource.rb

```diff
@@ -2,35 +2,59 @@ module Semian
   class ProtectedResource
     extend Forwardable
 
-    def_delegators :@resource, :destroy, :count, :semid, :tickets, :name
+    def_delegators :@bulkhead, :destroy, :count, :semid, :tickets, :name
     def_delegators :@circuit_breaker, :reset, :mark_failed, :mark_success, :request_allowed?,
                    :open?, :closed?, :half_open?
 
-    def initialize(resource, circuit_breaker)
-      @resource = resource
+    attr_reader :bulkhead, :circuit_breaker
+
+    def initialize(bulkhead, circuit_breaker)
+      @bulkhead = bulkhead
       @circuit_breaker = circuit_breaker
     end
 
     def destroy
-      @resource.destroy
-      @circuit_breaker.destroy
+      @bulkhead.destroy unless @bulkhead.nil?
+      @circuit_breaker.destroy unless @circuit_breaker.nil?
     end
 
     def acquire(timeout: nil, scope: nil, adapter: nil)
-      @circuit_breaker.acquire do
-        begin
-          @resource.acquire(timeout: timeout) do
-            Semian.notify(:success, self, scope, adapter)
+      if @circuit_breaker && !@bulkhead
+        acquire_circuit_breaker(scope, adapter) do
+          yield self
+        end
+      elsif @bulkhead && !@circuit_breaker
+        acquire_bulkhead(timeout, scope, adapter) do
+          yield self
+        end
+      else
+        acquire_circuit_breaker(scope, adapter) do
+          acquire_bulkhead(timeout, scope, adapter) do
             yield self
           end
-        rescue ::Semian::TimeoutError
-          Semian.notify(:busy, self, scope, adapter)
-          raise
         end
       end
+    end
```

Comment:
I moved the check for null in the `acquire_*` methods the problem is that if you define `&block` it expects the previous method parameters from my testing. I changed the method to be the same print as before.